### PR TITLE
GRASP-13023 Add retcode service

### PR DESCRIPTION
**Pull request by Kotek13**
Implemented retcode service.

Tested using [ConfigurationUploader.cpp in different scenarios](https://github.com/NoMagicAi/monomagic/pull/9922).

### CODE REVIEW

File: include/abb_librws/v1_0/rw/retcode.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+     * https://developercenter.robotstudio.com/api/rwsApi/rwretcode_get_return_codes.html
+     *
+     * \param client RWS client
+     * \param code code to be checked.
+     * 
+     * \return \a RWSResult containing the result.
+     *
+     * \throw \a RWSError if something goes wrong.
+     */
+    RWSResult getRetcodeInfo(RWSClient& client, int const& code);
```
Comment (mkatliar) [inline include/abb_librws/v1_0/rw/retcode.h:27]: `RWSResult` is used at a lower level. when you interact with `RWSClient` directly and don't interpret the result. Here we want to interpret the result as a structure describing a return code.

```suggestion
    RetcodeInto getRetcodeInfo(RWSClient& client, int const& code);
```

**Review by mkatliar — CHANGES_REQUESTED** [review]
Please add a structure for retcode info.

**Comment by mkatliar** [comment]
This PR seems to contain #50, so let's merge it after #50 is merged.

**Comment by Kotek13** [comment]
PR #50 merged.
Rebased.

File: include/abb_librws/v1_0/rw/retcode.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace abb :: rws :: v1_0 :: rw :: retcode
+{
+    /**
+     * \brief A structure for containing response of recode resource.
+     */
+    struct RetcodeInfo
+    {
+        int code;                   /** \brief The error code or success code as number */
+        std::string name;           /** \brief The error code or success code as string */
```
Comment (mkatliar) [inline include/abb_librws/v1_0/rw/retcode.h:21]: Is `name` just a string representation of `code`, like `"`123`", or something else? Please change the comment s.t. it is clear.

File: include/abb_librws/v1_0/rw/retcode.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        std::string severity;       /** \brief Success, Warning or Error */
+        std::string description;    /** \brief Short description of the error code */
+    };
+
+    /**
+     * \brief Error info containing RetcodeInfo struct.
+     */
+    using ErrorCodeInfo = boost::error_info<struct RetcodeInfoInfo, RetcodeInfo>;
```
Comment (mkatliar) [inline include/abb_librws/v1_0/rw/retcode.h:29]: We don't attach `ErrorCodeInfo` to exceptions that we throw from `abb_librws`, right? Then I don't think `ErrorCodeInfo` should be a part of `abb_librws`.

File: include/abb_librws/v1_0/rw/retcode.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * \brief A function for checking description of diagnostic code.
+     *
+     * https://developercenter.robotstudio.com/api/rwsApi/rwretcode_get_return_codes.html
+     *
+     * \param client RWS client
+     * \param code code to be checked.
+     * 
+     * \return \a RWSResult containing the result.
```
Comment (mkatliar) [inline include/abb_librws/v1_0/rw/retcode.h:39]: The return type is not `RWSResult`

File: include/abb_librws/v2_0/rw/retcode.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace abb :: rws :: v2_0 :: rw :: retcode
+{
+    /**
+     * \brief A structure for containing response of recode resource.
+     */
+    struct RetcodeInfo
+    {
+        int code;                   /** \brief The error code or success code as number */
+        std::string name;           /** \brief The error code or success code as string */
```
Comment (mkatliar) [inline include/abb_librws/v2_0/rw/retcode.h:21]: Same as for v1_0

File: include/abb_librws/v2_0/rw/retcode.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * \brief A function for checking description of diagnostic code.
+     *
+     * https://developercenter.robotstudio.com/api/rwsApi/rwretcode_get_return_codes.html
+     *
+     * \param client RWS client
+     * \param code code to be checked.
+     * 
+     * \return \a RWSResult containing the result.
```
Comment (mkatliar) [inline include/abb_librws/v2_0/rw/retcode.h:39]: The return type is not `RWSResult`

**Review by mkatliar — CHANGES_REQUESTED** [review]
Looks good, but some small changes are needed.

File: include/abb_librws/v1_0/rw/retcode.h (hunk 2 above)
Comment (Kotek13) [reply include/abb_librws/v1_0/rw/retcode.h:21]:
```suggestion
        std::string name;           /** \brief The name of error code or success code */
```

File: include/abb_librws/v1_0/rw/retcode.h (hunk 3 above)
Comment (Kotek13) [reply include/abb_librws/v1_0/rw/retcode.h:29]:
```suggestion
```

File: include/abb_librws/v1_0/rw/retcode.h (hunk 4 above)
Comment (Kotek13) [reply include/abb_librws/v1_0/rw/retcode.h:39]:
```suggestion
     * \return \a RetcodeInfo containing the result.
```

File: include/abb_librws/v2_0/rw/retcode.h (hunk 5 above)
Comment (Kotek13) [reply include/abb_librws/v2_0/rw/retcode.h:21]:
```suggestion
        std::string name;           /** \brief The name of error code or success code */
```

File: include/abb_librws/v2_0/rw/retcode.h (hunk 6 above)
Comment (Kotek13) [reply include/abb_librws/v2_0/rw/retcode.h:39]:
```suggestion
     * \return \a RetcodeInfo containing the result.
```

**Comment by Kotek13** [comment]
Moved `ErrorCodeInfo` into `RwsAdapter` and resolved other issues.

File: include/abb_librws/v2_0/rw/retcode.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+     * \brief A structure for containing response of recode resource.
+     */
+    struct RetcodeInfo
+    {
+        int code;                   /** \brief The error code or success code as number */
+        std::string name;           /** \brief The name of error code or success code */
+        std::string severity;       /** \brief Success, Warning or Error */
+        std::string description;    /** \brief Short description of the error code */
+    };
+
```
Comment (Kotek13) [inline include/abb_librws/v2_0/rw/retcode.h:25]:
```suggestion


```

File: include/abb_librws/v1_0/rw/retcode.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+     * https://developercenter.robotstudio.com/api/rwsApi/rwretcode_get_return_codes.html
+     *
+     * \param client RWS client
+     * \param code code to be checked.
+     * 
+     * \return \a RetcodeInfo containing the result.
+     *
+     * \throw \a RWSError if something goes wrong.
+     */
+    RetcodeInfo getRetcodeInfo(RWSClient& client, int const& code);
```
Comment (mkatliar) [inline include/abb_librws/v1_0/rw/retcode.h:39]: Passing an `int` by reference does not really make sense: you pass a pointer to `int`, which is bigger than the `int` itself, and also needs to be dereferenced. The compiler will optimize it out, but what is the purpose?
```suggestion
    RetcodeInfo getRetcodeInfo(RWSClient& client, int code);
```

File: include/abb_librws/v2_0/rw/retcode.h [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+     * https://developercenter.robotstudio.com/api/rwsApi/rwretcode_get_return_codes.html
+     *
+     * \param client RWS client
+     * \param code code to be checked.
+     * 
+     * \return \a RetcodeInfo containing the result.
+     *
+     * \throw \a RWSError if something goes wrong.
+     */
+    RetcodeInfo getRetcodeInfo(RWSClient& client, int const& code);
```
Comment (mkatliar) [inline include/abb_librws/v2_0/rw/retcode.h:39]:
```suggestion
    RetcodeInfo getRetcodeInfo(RWSClient& client, int code);
```

**Review by mkatliar — CHANGES_REQUESTED** [review]
No point in passing an `int` by reference, otherwise looks good.

File: src/v1_0/rw/retcode.cpp [hunk 10]
```diff
@@ -0,0 +1,18 @@
+#include <abb_librws/v1_0/rw/retcode.h>
+
+
+namespace abb :: rws :: v1_0 :: rw :: retcode
+{
+    RetcodeInfo getRetcodeInfo(RWSClient& client, int const& code)
```
Comment (Kotek13) [inline src/v1_0/rw/retcode.cpp:6]:
```suggestion
    RetcodeInfo getRetcodeInfo(RWSClient& client, int const code)
```

File: src/v1_0/rw/retcode.cpp [hunk 11]
```diff
@@ -0,0 +1,18 @@
+#include <abb_librws/v1_0/rw/retcode.h>
+
+
+namespace abb :: rws :: v1_0 :: rw :: retcode
+{
+    RetcodeInfo getRetcodeInfo(RWSClient& client, int const code)
```
Comment (Kotek13) [inline src/v1_0/rw/retcode.cpp:6]:
```suggestion
    RetcodeInfo getRetcodeInfo(RWSClient& client, int code)
```

File: src/v2_0/rw/retcode.cpp [hunk 12]
```diff
@@ -0,0 +1,18 @@
+#include <abb_librws/v2_0/rw/retcode.h>
+
+
+namespace abb :: rws :: v2_0 :: rw :: retcode
+{
+    RetcodeInfo getRetcodeInfo(RWSClient& client, int const code)
```
Comment (Kotek13) [inline src/v2_0/rw/retcode.cpp:6]:
```suggestion
    RetcodeInfo getRetcodeInfo(RWSClient& client, int code)
```

**Pull request merged**